### ISSUE TITLE
Dynamic Prompts doesn't play nice with other prompt-modifying extensions

### ISSUE THREAD

**Issue by artyfacialintelagent** [adieyal/sd-dynamic-prompts#543]
I am the author of [CloneCleaner](https://github.com/artyfacialintelagent/CloneCleaner), an extension that uses wildcards behind the scenes to achieve more image variability for different seeds. Several of my users have reported a conflict with Dynamic Prompts, see https://github.com/artyfacialintelagent/CloneCleaner/issues/1.

I've diagnosed the problem and have determined that the underlying cause is that Dynamic Prompts only reads the first prompt of a batch (i.e. `p.all_prompts[0]`), applies that prompt to all images of a batch and then modifies those prompts. Since DP appears to run after CloneCleaner (do extensions always run alphabetically?), DP overwrites the prompt changes in the rest of `p.all_prompts` that CloneCleaner has made.

What DP should do to play nice is instead to read the entire array `p.all_prompts` and modify the prompts for individual seeds based on the wildcards in `p.all_prompt`. If you make this change I believe our extensions should be compatible - I'd appreciate that!

**Comment by GalaxyTimeMachine** [comment]
Same problem here.

**Comment by adieyal** [comment]
@artyfacialintelagent - do you modify `all_hr_prompts` as well? Handling the various prompts is surprisingly not as trivial as you would expect. In addition, Dynamic Prompts maintains state in between generations. For example, with combinatorial generation, we generate all the combinations for a particular template, e.g. a `{red|green|blue} ball` will produce:

```
a red ball
a green ball
a blue ball
```

If CloneCleaner creates multiple prompts, then we would need to run combinatorial for each one. I'm not sure that the two extensions are compatible, at least if CloneCleaner runs before Dynamic Prompts. Let me know if you have any ideas.

**Comment by artyfacialintelagent** [comment]
No, I only modify `p.all_prompts[i]`, nothing else.

Judging from the comments I've seen, I think the main use case for running CloneCleaner and Dynamic Prompts at the same time is to get additional custom wildcards in addition to CloneCleaner's automatic ones. My guess is that the people who use combinatorial generation want to enumerate all prompt combinations deterministically, and those users probably don't want the prompt randomization that my extension does. So I don't think we need to worry about doing combinatorial generation over the different prompts I'm generating.

So as far as I'm concerned getting wildcards to work would be enough to make most people happy. CloneCleaner only appends stuff to the beginning or end of the prompt - or rather each prompt instance `p.all_prompts[i]`. I've been careful to leave the rest of the prompt unchanged, so any `__wildcards__` are still unmodified. I scanned through your source code when I opened this issue and concluded that the code [here](https://github.com/adieyal/sd-dynamic-prompts/blob/fa49646b61f86930885b86f34dea0b089dfac168/sd_dynamic_prompts/generator_builder.py#L217) and [here](https://github.com/adieyal/sd-dynamic-prompts/blob/fa49646b61f86930885b86f34dea0b089dfac168/sd_dynamic_prompts/dynamic_prompting.py#L60) only reads `p.all_prompts[0]` and ignores the other indices. That's the root of the problem I believe, but fixing it will require changes elsewhere as well. I don't think the state you maintain between generations plays into this, but I could be wrong.

**Comment by adieyal** [comment]
Having given it some thought, I don't think that Dynamic Prompts should work in a way that would address your problem. The design of the extension is such that it takes a template as input and generates new prompts from it.  There are too many cases where state is needed for it to handle multiple incoming prompts. Combinatorial mode is one example, the [cyclical operator](https://github.com/adieyal/sd-dynamic-prompts/blob/main/docs/SYNTAX.md#cyclical-sampler) is another. Another problematic feature is [Jinja2 templates](https://github.com/adieyal/sd-dynamic-prompts/blob/main/jinja2.md) which can generate multiple new prompts. 

@h43lb1t0's work-around is the simplest way to handle it. An alternative is to perhaps set-up a hook so that Dynamic Prompts calls CloneCleaner once it has finished creating new prompts. Neither solution is 100% perfect. Perhaps the best solution is to implement a dependency manager in auto1111.

**Comment by artyfacialintelagent** [comment]
Well, the design of your extension is up to you. But this issue isn't really a feature request - frankly it's a bug. It didn't manifest when you were the only player in the prompt-modifying field but now I'm here too, and it's a safe bet others will enter soon. They will experience the same conflict with your extension that I did.

By interpreting `p.all_prompts[0]` as the template and overwriting `p.all_prompts[i]` based on that template, Dynamic Prompts by design destroys every prompt in a batch except the first. Basically you're assuming that `p.all_prompts[i]` are all identical. That may be true in vanilla Auto1111 but it doesn't hold when other extensions are involved. 

We might be able to avoid this problem for now by hacking the extension run order, as in the rename suggestion and your hook idea. But then I think I'd rather submit a PR upstream to Auto1111 that implements custom extension order once and for all. Issue [#8011](https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/8011) requests this, so that would help others too.

But the underlying problem is inherent in the single-template approach, and I think you'll need to refactor your extension sooner or later. Somehow you need to figure out a way to make local changes to individual prompts - even if they have state and depend on each other - without clobbering parts of those prompts outside your template words.

**Comment by adieyal** [comment]
It's definitely not a bug but a design decision. In a nutshell, the extension does the following:

1. User provides prompt template with wildcards, etc
2. Dynamic Prompts uses that template to generate new prompts according to its DSL and certain rules.

Currently  this works as `template -> multiple prompts.`

If I understand correctly, your extension does the same.  The change that would need to be made for both extensions to play nicely is:
```
user prompt
  ─> CloneCleaner prompt 1
     ─> Dynamic Prompts prompt 1a
     ─> Dynamic Prompts prompt 1b
  ─> CloneCleaner prompt 2
     ─> Dynamic Prompts prompt 2a
     ─> Dynamic Prompts prompt 2b
```

The user asked for 2 images but they get 4. In fact, it's not even clear how many prompts to create. If `batch_size` is 1 and `batch_count`  is 4, do we produce four CloneCleaner prompts and four DynamicPrompts for each CloneCleaner prompt?

**Comment by artyfacialintelagent** [comment]
No, several people have reported that both extensions work when CloneCleaner runs after Dynamic Prompts. They work correctly because CloneCleaner first reads all the modified prompts in `p.all_prompts[i]` before writing back to them. So any modifications Dynamic Prompts has already made get passed through unchanged - or rather, with some added variability from the randomized prompt inserts.

As for your last question - if I understand it correctly, you wonder what Dynamic Prompts should do when the input prompts vary over the batch (as in the output from CloneCleaner). But as I see it, that shouldn't be your concern. DP should only edit what's inside the special syntax delimiters that DP supports, e.g. double underscores, curly brackets, etc. You wrote an excellent parser to identify those parts, and your job is to do fancy word replacements and combinatorics in those parts. But anything outside those DP sections should just be passed through unchanged, and should be stitched together with your replacements at the end. At least that's my view. So that's why I called it a bug when you changed stuff outside your delimiters.

I think our whole misunderstanding stems from this. You appear to view the entire prompt as your input template, and with that view variable inputs over a batch are a massive headache. That's understandable. But in my view, CloneCleaner shouldn't affect you or be affected by you at all since it passes through the DP syntax delimiters unchanged. It's kind of a global/local thing.

But I understand that you wrote DP to manipulate a single input template and changing its behavior now would require a pretty significant refactor. So that's why I'm going to see if we can fix this by tweaking the extension run order. But once again, the next prompt-editing extension to come along will probably also conflict with DP for the same reason.
